Implement basic blocks & threaded code

VasiliyMatr opened the pull request:
* Move status codes to common module
* Add routines for load & store ops
* Add bb module
* Add integration tests for simulator module
* Implement threaded code
* A few other fixes

VasiliyMatr committed:
[all] Move status codes to common module

- Separate status enums are replaced with aggregate status enum in common module

VasiliyMatr committed:
[simulator] Add routines for load & store ops

VasiliyMatr committed:
[bb] Add bb module

- Implement bb::Bb class for decoded instrs collecting

VasiliyMatr committed:
[simulator] Integrate bb::Bb

- Use basic blocks in simulator::Simulator
- Add icount
- A few changes in sim_app

VasiliyMatr committed:
[simulator] Add new instrs & fix load/store sim

- Add support for: ADDI, SLTI, SLTIU, ANDI, ORI, XORI
- Fix load/store simulation bugs - offset is used now

VasiliyMatr committed:
[simulator] Add integration tests

VasiliyMatr committed:
[cmake][riscv] Move risc-v.yaml to proj root

VasiliyMatr committed:
[memory] Minor fixes

- Fix MMU PTE validity check
- A few other fixes

VasiliyMatr committed:
[simulator][instr][bb] Impl threaded code draft

VasiliyMatr committed:
[instr][simulator] Fix tests after API changes

- Fix instr and simulator modules tests

TheRedHotHabanero commented on bb/include/sim/bb.hpp:
```diff
@@ -0,0 +1,93 @@
+            return false;
+        }
+
+        SIM_ASSERT(0);
```
This line is unreachable

VasiliyMatr replied:
And that's why I put an assertion there

TheRedHotHabanero commented on bb/include/sim/bb.hpp:
```diff
@@ -0,0 +1,93 @@
+        // Decode instrs starting from bb_virt_addr
+        m_virt_addr = bb_virt_addr;
+
+        for (size_t i = 0; i < MAX_SIZE - 1; ++i) {
```
Is it correct? It will iterate [0...14]

VasiliyMatr replied:
Yes, it is correct. Since there is status instruction insertion after the cycle. This instruction terminates bb

TheRedHotHabanero commented on csr/src/csr.cpp:
```diff
@@ -1,7 +1,18 @@
+#include "sim/common.hpp"
+#include "sim/csr/idx.gen.hpp"
+#include "sim/csr/value.gen.hpp"
```
Common style for headers?

VasiliyMatr replied:
Fixed

TheRedHotHabanero commented on instr/CMakeLists.txt:
```diff
@@ -7,7 +7,6 @@ PUBLIC
     sim::common
 )
 
-set(RISCV_YAML ${CMAKE_CURRENT_SOURCE_DIR}/risc-v.yaml)
```
Why did you delete this line?

VasiliyMatr replied:
Because risc-v.yaml is moved to the project root

TheRedHotHabanero commented on memory/include/sim/memory/mmu64.hpp:
```diff
@@ -1,6 +1,7 @@
 #ifndef INCL_MEMORY_MMU64_HPP
 #define INCL_MEMORY_MMU64_HPP
 
+#include "sim/common.hpp"
```
Common style for headers?

VasiliyMatr replied:
Fixed

TheRedHotHabanero commented on sim_app/src/sim_app.cpp:
```diff
@@ -1,3 +1,4 @@
+#include "sim/common.hpp"
```
Common include style

VasiliyMatr replied:
Fixed

TheRedHotHabanero commented on simulator/include/sim/simulator.hpp:
```diff
@@ -1,37 +1,186 @@
 #ifndef INCL_SIM_SIMULATOR_HPP
 #define INCL_SIM_SIMULATOR_HPP
 
+#include "sim/instr/instr_id.gen.hpp"
```
Style

VasiliyMatr replied:
Fixed

TheRedHotHabanero commented on simulator/tests/src/test_simulator.cpp:
```diff
@@ -0,0 +1,125 @@
+#include "sim/common.hpp"
```
Style

VasiliyMatr replied:
Fixed

VasiliyMatr committed:
[all] Replace unneeded includes

Note: clangd extension inserts includes automatically when some symbols are used in code.
Such behaviour is inconvenient. And can be turned off with option: '-header-insertion=never'
